File: 프로그래머스/2/388352. 비밀 코드 해독/비밀 코드 해독.py
def is_valid_case(queue, answer, current_case):
    for q, ans in zip(queue, answer):
        if sum(c in q for c in current_case) != ans:
            return False
        
    return True


def solution(n, q, ans):
    def dfs(x, current_case, result):
        if len(current_case) == 5:
            if is_valid_case(q, ans, current_case):
                result.append(current_case)
            return
        
        if x <= n:
            dfs(x + 1, current_case, result)
            dfs(x + 1, current_case + [x], result)
    
    result = []
    dfs(1, [], result)
    return len(result)

File: 프로그래머스/2/388352. 비밀 코드 해독/비밀 코드 해독.java
import java.util.*;

class Solution {
    
    private static final List<Set<Integer>> queue = new ArrayList<>();
    private static final List<List<Integer>> result = new ArrayList<>();
    
    private static int N;
    private static int[] answer;

    private static boolean isValidCase(List<Integer> currentCase) {
        
        for (int i = 0; i < queue.size(); i++) {
            Set<Integer> q = queue.get(i);
            int ans = answer[i];
            int count = 0;
            
            for (int c : currentCase) {
                if (q.contains(c)) count++;
            }
            
            if (count != ans) {
                return false;
            }
        }
        
        return true;
    }

    private static void dfs(int x, List<Integer> currentCase) {
        if (currentCase.size() == 5) {
            if (isValidCase(currentCase)) {
                result.add(new ArrayList<>(currentCase));
            }
            return;
        }
        
        if (x <= N) {
            dfs(x + 1, currentCase);
            currentCase.add(x);
            dfs(x + 1, currentCase);
            currentCase.remove(currentCase.size() - 1);
        }
    }

    public int solution(int n, int[][] qArr, int[] ans) {
        
        N = n;
        answer = ans;
        
        for (int[] arr : qArr) {
            Set<Integer> s = new HashSet<>();
            for (int v : arr) {
                s.add(v);
            }
            queue.add(s);
        }
        
        dfs(1, new ArrayList<>());
        
        return result.size();
    }
}